### PR TITLE
Packages: reduce number of GET requests to sync packages

### PR DESCRIPTION
**Pull request by olafurpg**
Previously, we sent a large number of requests to sync packages. My rough estimate is that we sent 3 GET requests for every tracked package+version combination. After this PR, we send only one GET request for every tracked package+version, and for npm we additionally send one GET request per package (regardless of number of versions) in order to fetch the package description.

This PR also includes a few more renames that should have been part of #37999

- rename "go modules" into "go packages" (for consistency with all other package implementations, including "jvm packages" even if JVM `package` is similar to Go `package`)
- rename `PackageVersion` into `VersionedPackage`, fixes https://github.com/sourcegraph/sourcegraph/issues/38009

Fixes https://github.com/sourcegraph/sourcegraph/issues/37990 

## Test plan

See the CI go green. I manually tested the changes locally by syncing 120k crate versions in less than one hour. For comparison, it took us ~30hr on sourcegraph.com to sync the same number of crate versions.

### CODE REVIEW

**Review by olafurpg — COMMENTED** [review]
The large diff in the golden test for npm is working as expected. Previously, we picked the description of a random version (the one that appeared first in the site-admin configuration of the `lsif_dependency_repos` table). Now, we pull the description of the package itself, which should be more accurate. The large diff origins from the fact that `react` has a lot of releases and the the summary of react (https://registry.npmjs.org/react) contains metadata about all published releases. I'm gonna update the golden tests to use a different  npm package with fewer releases to reduce the size of the diff.

**Review by Strum355 — COMMENTED** [review]
Some clarifications and questions that are adjacent to the review comments:

--- 
`Previously, we sent a large number of requests to sync packages...3 GET requests for every tracked package+version combination`

To confirm, these 3 requests were the following?
1. When syncing an external service (where `*PackagesSource.ListRepos` is called)
2. When syncing an individual repo, triggered by a repo lookup (where `*PackageSource.GetRepo` is called
3. When fetching a package in gitserver, `*vcsPackagesSyncer.Fetch`

---
`After this PR, we send only one GET request for every tracked package+version`

This one request is on the gitserver side in `*vcsPackagesSyncer.Fetch`, correct?

---

The changes here, as I understood, will result in potentially empty repos being created for packages that seem valid but cannot be resolved for whatever reason. 
I have some reservations on doing this, for site-administration reasons on having a potentially large number of (forever?) empty repositories polluting the site-admin UI. I dont feel super strongly about this, so Id have no major problems with this behaviour, it's only something that I feel is subpar user experience. 
On the flip side, it may be preferable to see these repos to surface syncing errors, if thats possible to do :shrug:

**Comment by Strum355** [comment]
I see this PR says it fixes https://github.com/sourcegraph/sourcegraph/issues/37990, but I feel 37990 is a different issue. After this PR is still merged, we'll still be making requests for every tracked package on sync, but my understanding of the issue is that tracked package+version combinations should only be requested once in their lifetime in the instance, as opposed to in every sync

File: cmd/gitserver/server/vcs_dependencies_syncer.go [hunk 1]
```diff
@@ -41,15 +41,21 @@ type packagesSource interface {
...
+	// field from the site-admin interface.
+	ParseVersionedPackageFromConfiguration(dep string) (reposource.VersionedPackage, error)
+	// ParsePackageFromRepoName parses a Sourcegraph repository name of the package.
 	ParsePackageFromRepoName(repoName string) (reposource.Package, error)
 }
 
+type packagesDownloadSource interface {
+	// GetPackage sends a request to the package host to get metadata about this package, like the description.
+	GetPackage(ctx context.Context, name string) (reposource.Package, error)
+}
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply cmd/gitserver/server/vcs_dependencies_syncer.go:57]: Good point. I opened an issue to followup on this.

[the remark this answers is not in the document]
Comment (olafurpg) [reply cmd/gitserver/server/vcs_dependencies_syncer.go:57]: https://github.com/sourcegraph/sourcegraph/issues/38571

**Comment by olafurpg** [comment]
Thank you for the review!

> I see this PR says it fixes https://github.com/sourcegraph/sourcegraph/issues/37990, but I feel 37990 is a different issue. After this PR is still merged, we'll still be making requests for every tracked package on sync, but my understanding of the issue is that tracked package+version combinations should only be requested once in their lifetime in the instance, as opposed to in every sync

I added a second commit to this PR that stops sending GET requests for all tracked versions during gitserver syncs. After this PR is merged, we should only send one GET request in total per version.

File: internal/repos/packages.go [hunk 2]
```diff
@@ -39,20 +53,18 @@ func (s *PackagesSource) ListRepos(ctx context.Context, results chan SourceResul
 		return
 	}
 
-	var mu sync.Mutex
-	set := make(map[string]struct{})
+	handledPackages := make(map[string]struct{})
 
 	for _, dep := range deps {
-		_, err := s.src.Get(ctx, dep.PackageSyntax(), dep.PackageVersion())
```
Comment (olafurpg) [inline internal/repos/packages.go:46]: We eliminate one GET request here for every tracked version. Instead, we only perform one GET request per **package** (not version) for code hosts that support the description field, which is currently only the npm code host.

File: internal/repos/packages.go [hunk 3]
```diff
@@ -87,32 +99,31 @@ func (s *PackagesSource) ListRepos(ctx context.Context, results chan SourceResul
...
+		for _, depRepo := range depReposToHandle {
 			if err := sem.Acquire(ctx, 1); err != nil {
 				return
 			}
-
 			depRepo := depRepo
 			g.Go(func() error {
 				defer sem.Release(1)
-
-				dep, err := s.src.Get(ctx, depRepo.Name, depRepo.Version)
```
Comment (olafurpg) [inline internal/repos/packages.go:99]: Same as above, we no longer send GET requests for all versions, we only send it once per package (not version) for code hosts that support the description field (only npm at the moment)

**Review by olafurpg — COMMENTED** [review]
The diff is unnecessarily complicated to review because it mixes a refactoring (renaming symbols) with non-trivial code changes (reduce get requests). I left comments on the critical parts where we remove GET requests. I can open a separate PR for the renames if there's strong desire for that, I can look into it tomorrow. However, I'd prefer to land this PR asap so that we stop sending so many requests.

**Review by Strum355 — APPROVED** [review]
Gave another brief look through, LGTM. If someone else wants to give it another clean-slate read through to catch anything out, thatd be cool beans

I've changed my mind on the empty repos part, so long that the repo is marked as "Not yet cloned" in site admin instead of looking like a successfully cloned but empty repo :sunglasses:

**Comment by olafurpg** [comment]
@Strum355 you might be interested in the updated test suite that should do an OK job to document the new behavior.

**Pull request merged**